AUNF Mini-Canister Example Case (#410)
idaholab/TMAP8: doc/content/examples/mini_canister/index.md

Reviewer: I suggest adding the visualization from the manuscript for the two cases.

Author: Added the cylinder graphic from the manuscript.

# Mini-Canister Hydrogen Transport

> TMAP8 is used to model hydrogen transport and permeation through an aluminum-clad used nuclear fuel (AUNF) mini-canister storage device from Savannah River National Laboratory (SRNL) [!citep](d'entremont2024aunfminicanister). The mini-canisters house irradiated AUNF assemblies where gamma and neutron radiation from the fuel drives radiolytic decomposition of water, generating H$_2$ gas. Over time, this hydrogen will dissociate and diffuse through the surrounding 304 stainless steel wall, raising concern for potential accumulation. This example demonstrates how TMAP8 can model these processes through two distinct input files:

Reviewer: ```suggestion
TMAP8 is used to model hydrogen transport and permeation through an aluminum-clad used nuclear fuel (AUNF) mini-canister storage device from Savannah River National Laboratory (SRNL) [!citep](d'entremont2024aunfminicanister). The mini-canisters house irradiated AUNF assemblies where gamma and neutron radiation from the fuel drives radiolytic decomposition of water, generating H$_2$ gas. Over time, this hydrogen will dissociate and diffuse through the surrounding 304 stainless steel wall, raising concern for potential accumulation. This example demonstrates how TMAP8 can model these processes through two distinct input files with varying degrees of fidelity:
```

!table id=tab:geometry caption=Canister geometry.
| Parameter | Value | Units |
| --- | --- | --- |
| Inner radius, $\mathbf{r_i}$ | $\mathbf{35.94}$ | mm |
| Steel wall thickness, $\mathbf{\delta_s}$ | $\mathbf{2.16}$ | mm |
| Outer radius, $\mathbf{r_o} = \mathbf{r_i} + \mathbf{\delta_s}$ | $\mathbf{38.10}$ | mm |
| Canister height, $\mathbf{h}$ | $\mathbf{179.3}$ | mm |
| Internal gas volume, $\mathbf{V_g} = \pi\mathbf{r_i}^2\mathbf{h}$ | $\approx \mathbf{727,400}$ | mm$^3$ |

In [steel_only.i], the 1D mesh spans only the steel wall, from $r_i$ to $r_o$, using a single [GeneratedMeshGenerator.md] with 1,500 elements (subdomain 1). In [gas_steel.i], a [CartesianMeshGenerator.md] produces two adjacent blocks: the gas block (subdomain 0) from $r = 0$ to $r = r_i$ with 250 elements, and the steel block (subdomain 1) from $r = r_i$ to $r = r_o$ with 1,500 elements. Two [SideSetsBetweenSubdomainsGenerator.md] steps then create the named interface sidesets `interface_gas_to_steel` and `interface_steel_to_gas` that are required for the [InterfaceSorption.md] kernel.

> 
> In [steel_only.i], the mesh is defined as:
> !listing test/tests/mini_canister/steel_only.i link=false block=Mesh

Reviewer: ```suggestion
In [steel_only.i], the mesh is defined as:

!listing test/tests/mini_canister/steel_only.i link=false block=Mesh
```

> 
> In [gas_steel.i], the mesh is defined as:
> !listing test/tests/mini_canister/gas_steel.i link=false block=Mesh

Reviewer: ```suggestion
In [gas_steel.i], the mesh is defined as:

!listing test/tests/mini_canister/gas_steel.i link=false block=Mesh
```

\begin{equation} \label{eq:sieverts_bc}
C_s(r_i, t) = 2 K_s(T) \sqrt{P},
\end{equation}

where the factor of 2 converts from molecular H$_2$ equilibrium to atomic H concentration, and the solubility follows an Arrhenius dependence:

\begin{equation} \label{eq:solubility}
K_s(T) = K_{s,0} \exp\!\left( - \frac{E_{K}}{RT} \right).
\end{equation}

!style halign=left
The H$_2$ partial pressure $P$ is provided by a function selected via the `pressure_function` input parameter. Three options are available: (1) a constant value (`constant_pressure`), (2) a [TimeRampFunction.md] from zero to the constant value over 3 hours (`time_ramp_pressure`), or (3) a power-law fit to experimental SRNL pressure data (`SRNL_pressure_data_fun`):

\begin{equation} \label{eq:srnl_pressure}
P_{\text{SRNL}}(t) = 376.7588 \, t^{0.6177} \quad [\text{Pa}].
\end{equation}

The default configuration uses the constant pressure option:

\begin{equation} \label{eq:assumed_pressure}
P = 0.10 \times 24 \text{ psi} \approx 16{,}547 \text{ Pa},

> \end{equation}

Reviewer: While you have investigated all these options, there is no need to discuss them all here. Focus on the one you have used. 

Also, there are virtually an infinite number of options available, so stating there are just 3 is a bit misleading.

Author: Constant pressure is used for verification with diffusion length and SRNL pressure is used for validation against data, so I will include both of those. I'll remove discussion on the time ramping option.

which assumes 10% of the 24 psi He-backfilled canister pressure is attributable to H$_2$ [!citep](d'entremont2024aunfminicanister,hlushko2024aunf).

At the outer steel surface ($r = r_o$), hydrogen is released to the ambient environment, and the concentration is set to zero by a [DirichletBC.md] (defined in [mini_canister_base.i]):

\begin{equation} \label{eq:outer_bc}
C_s(r_o, t) = 0.
\end{equation}

### Solver

!style halign=left
Because the steel-only problem is linear (constant diffusivity, linear Sieverts' BC), [steel_only.i] uses `solve_type = LINEAR` for a direct LU factorization at each timestep.

> ### Model Parameters

Reviewer: ```suggestion
### Model Parameters and Simulation Conditions
```

> !table id=tab:steel_only_params caption=Steel-only model parameters.

Reviewer: ```suggestion
!table id=tab:steel_only_params caption=Steel-only model parameters and simulation conditions.
```
While the diffusivity and solubilities are model parameters, the temperature and the pressure are conditions and not model parameters. 

Let me know if the distinction does not make sense and we can discuss this.

| Parameter | Description | Value | Units | Reference |
| --- | --- | --- | --- | --- |
| $\mathbf{D_{s,0}}$ | Diffusivity pre-exponential factor | $\mathbf{0.20 \times 10^{-6}}$ | m$^2$ s$^{-1}$ | [!cite](san_marchi2012hydrogensteel) |
| $\mathbf{E_{D}}$ | Diffusivity activation energy | $\mathbf{49.3}$ | kJ mol$^{-1}$ | [!cite](san_marchi2012hydrogensteel) |
| $\mathbf{K_{s,0}}$ | Solubility pre-exponential factor | $\mathbf{266 \times 10^{-6}}$ | $\mathrm{\mu}$mol mm$^{-3}$ Pa$^{-1/2}$ | [!cite](san_marchi2012hydrogensteel) |
| $\mathbf{E_{K}}$ | Solubility activation energy | $\mathbf{6.86}$ | kJ mol$^{-1}$ | [!cite](san_marchi2012hydrogensteel) |
| $\mathbf{T}$ | Temperature | $\mathbf{313.15}$ | K | [!cite](d'entremont2024aunfminicanister) |
| $\mathbf{P}$ | Assumed H$_2$ partial pressure | $\approx\mathbf{16,547}$ | Pa | [!cite](d'entremont2024aunfminicanister) |

### Results

#### Diffusion Front Verification

!style halign=left

> For a semi-infinite slab with a constant-concentration boundary condition, the diffusion front advances as $\ell(t) = \sqrt{\pi D_s t}$, defined here as the $r$-intercept of the tangent line to the concentration profile at the inner surface. This analytical result provides a straightforward check that the numerical diffusion is correctly implemented. The simulated diffusion front is computed via the `simulated_diffusion_length` postprocessor as the $r$-intercept of the tangent line using the interface concentration and gradient. [fig:diffusion_length] shows the simulated diffusion front length, using a constant pressure, compared to the analytical expression.

Reviewer: ```suggestion
For a semi-infinite slab with a constant-concentration boundary condition, the diffusion front advances as $\ell(t) = \sqrt{\pi D_s t}$, defined here as the $r$-intercept of the tangent line to the concentration profile at the inner surface. This analytical result provides a straightforward check that the numerical diffusion is correctly implemented. The simulated diffusion front is computed via the `simulated_diffusion_length` postprocessor as the $r$-intercept of the tangent line using the interface concentration and gradient. [fig:diffusion_length] shows the simulated diffusion front length, using a constant pressure, compared to the analytical expression, which it matches.
```

!media comparison_mini_canister.py

>   image_name=diffusion_length.png

Reviewer: The two curves overlap perfectly. This is great for the verification, but not good to be able to actually see the curves. Please check out the verification cases to see the colors and ls option we usually use for the verification cases to plot the analytical solution vs the TMAP8 calculations.

  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:diffusion_length
  caption=Comparison of the simulated and analytical ($\sqrt{\pi D_s t}$) diffusion front length in the steel wall over 0.25 years.

> 

Reviewer: ```suggestion

!listing test/tests/mini_canister/gas_steel.i link=false block=Postprocessors/exact_diffusion_length

!listing test/tests/mini_canister/gas_steel.i link=false block=Postprocessors/simulated_diffusion_length

```

#### Conservation of Mass

!style halign=left
As an internal consistency check, the total hydrogen mass integrated over the steel domain (`annular_cylinder_total_mass_steel`) is compared against the time-integrated net diffusive flux across the inner and outer boundaries (`annular_cylinder_time_integrated_flux`). Both integrals are weighted for the axisymmetric cylindrical geometry and then scaled by the canister height $h$ to represent 3D mass in $\mathrm{\mu}$mol H. [fig:steel_conservation] shows the two quantities track closely throughout the simulation, confirming mass conservation.

!media comparison_mini_canister.py

>   image_name=steel_only_conservation_of_mass.png

Reviewer: Plot instead the ratio of total mass over the total internal mass, minus 1. This way the results should stay around 0.
Note not to do this math in the postprocessors but in the python script, because otherwise it would create a very sensitive pps.

Author: Why do we need this change? What is the benefit of the results staying around 0?

Verifying formula: $\frac{M_s - M_g}{M_g}-1$

Reviewer: I think the limit for the mass conservation should be based on the ratio, not the absolute value.
You get a better idea of the magnitude of the error with respect to the value.
Because if you start with an absolute value of 1e30, then variations of 1e20 would potentially be good.
But if you use the ratio, then you can say the variations are within a certain percentage, which helps you compare accuracy across different conditions and total amounts. 

The right formula is: $\frac{M(t)}{M_{init}}-1 = \frac{M(t)-M_{init}}{M_{init}}$
with M(t) the mass as a function of time, and M_{init} the initial mass.

Author: Seems that the general idea is to produce a relative ratio of how much mass is unaccounted for by conservation of mass. This metric would also remove many of the postprocessors present, as the accumulated flux on the boundaries is unneeded. I am fine with this and think it is an improvement.

So $M_{init}$ would be representative of the internal mass in the gas-phase from your first comment, or the initial mass from your most recent comment? The inital mass is 0 in our case so this metric would need to be modified. 

Also, the steel-only model just measures total mass in the steel, so $M_{init}$, if this is the internal gas-phase mass, is not measured. Should we use ideal gas law to approximate this as InterfaceSorption does in the full model?

Reviewer: Sorry for the lack of clarity, I apparently wasn't thinking straight during my last post. 

For now, you compare two values in the mass conservation plots and want to show that they are the same.
My point is that the best way to do that is to do the ratio of the two, minus one. 
In your case, you can do 
$\frac{\text{integrated flux}}{\text{total mass}}$ - 1
and do the equivalent for both cases. 

I hope that helps.

Author: Implemented the formula and attached the results. This metric is relative and is quite sensitive in the early timesteps, though we see quick convergence to a reasonable percentage difference in COM for both models. Some of the other verification cases only consider the time after a certain point for the RMSPE calculations as to not skew that metric, so perhaps I could just plot these graphs after 1 day simulation time? Let me know what you think. @lin-yang-ly have any suggestions?

<img width="1711" height="1456" alt="Image" src="https://github.com/user-attachments/assets/5f6ad395-5b61-4085-930d-84aa9fa6b019" />
<img width="1711" height="1456" alt="Image" src="https://github.com/user-attachments/assets/6ff4a2bc-3d01-4fef-84d9-bd68ac89680e" />

Reviewer: The reason why they only look at RMSPE after some time is because the earlier values are 0 or unphysical. Here, the situation is different: mass conservation is not well imposed at first. I recommend these few things:
- make sure the pps execute at timestep_final and not at initial. That should avoid initial values that are set to 0 and cause issues for the first timestep. 
- do a dt convergence study to make sure your initial timestep is small enough. This is the whole point of the mass conservation study: to make sure you are solving the system in an accurate way.

### Hydrogen Generation Source Term

!style halign=left
The cumulative radiolytic H$_2$ yield is modeled using a power-law fit to the SRNL experimental data [!citep](d'entremont2024aunfminicanister):

\begin{equation} \label{eq:H2_yield}
N(t) = 69.7055 \, t^{0.6808} \quad [\mathrm{\mu}\text{mol}],
\end{equation}

where $t$ is time in days, calibrated assuming a Co-60 irradiator dose rate of approximately 124 Gy/min [!citep](d'entremont2024aunfminicanister). The volumetric source term in [eq:gas_diffusion] is obtained by differentiating [eq:H2_yield] with respect to time and normalizing by the gas volume:

\begin{equation} \label{eq:source_term}
S(t) = \frac{1}{V_g} \frac{dN}{dt} = \frac{69.7055 \times 0.6808}{V_g} \, t^{0.6808 - 1} \quad [\mathrm{\mu}\text{mol mm}^{-3}\text{ day}^{-1}].

> \end{equation}

Reviewer: You should elaborate on how the exact terms and coefficients are derived.

Author: Added lines near the pressure and cumulative yield fit specifying how these were calculated.

\begin{equation} \label{eq:interface_sieverts}
C_s(r_i, t) = 2 K_s(T) \sqrt{C_g(r_i, t) \, R \, T},
\end{equation}

The `unit_scale_neighbor` parameter is set to $10^3$ to correct for the unit mismatch between $C_g$ in $\mathrm{\mu}$mol mm$^{-3}$ and the ideal gas constant used internally by [InterfaceSorption.md], which draws $R$ in J K$^{-1}$ mol$^{-1}$ from the MOOSE [TMAP8PhysicalConstants.md] namespace. Converting $C_g$ from $\mathrm{\mu}$mol mm$^{-3}$ to mol m$^{-3}$ introduces a combined factor of $10^{-6}$ (mol/$\mathrm{\mu}$mol) $\times$ $10^{9}$ (mm$^3$/m$^3$) $= 10^3$, which is supplied via `unit_scale_neighbor`.

### Solver

!style halign=left
Because the gas-steel problem is nonlinear (the interface couples $C_g$ and $C_s$ through a square-root relationship), [gas_steel.i] uses `solve_type = Newton` with `line_search = NONE`.

### Model Parameters

> 

Reviewer: ```suggestion

[tab:gas_steel_params] lists the parameters used in the gas-steel model. 

```

| Parameter | Description | Value | Units | Reference |
| --- | --- | --- | --- | --- |
| $\mathbf{D_g}$ | H$_2$ diffusivity in He | $\mathbf{2.7}$ | cm$^2$ s$^{-1}$ | [!cite](middha2002hydrogenhelium) |
| $\mathbf{V_g} = \pi\mathbf{r_i}^2\mathbf{h}$ | Internal gas volume | $\approx\mathbf{727,400}$ | mm$^3$ | Computed |

### Results

#### Partial Pressure Validation

!style halign=left
[fig:partial_pressure] compares the simulated H$_2$ partial pressure at the gas-steel interface against pressure measurements from the SRNL irradiation experiment [!citep](d'entremont2024aunfminicanister). The partial pressure is computed from the gas-phase concentration at the interface using the ideal gas law via the `H_partial_pressure_interface` postprocessor.

!media comparison_mini_canister.py
  image_name=partial_pressure_comparison.png
  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:partial_pressure
  caption=Comparison of TMAP8 simulated H$_2$ partial pressure at the gas-steel interface against SRNL experimental measurements.

#### Gas-Phase Hydrogen Yield Validation

!style halign=left

> [fig:gas_yield] compares the total atomic hydrogen mass in the gas phase (`inner_cylinder_total_mass_gas`) against the cumulative H$_2$ yield measured by SRNL [!citep](d'entremont2024aunfminicanister). Agreement between the simulation and experiment reflects the accuracy of the power-law source model ([eq:H2_yield]). It is important to note that we are comparing against data that is also fed into the model. Future work will include a more complex generation model independent of this data.

Reviewer: As stated above, make it explicit how the source term is calculated, so that this discussion here does not come as a surprise to the reader.

!media comparison_mini_canister.py

>   image_name=gas_phase_validation.png

Reviewer: The `H_2` in the y label should be rendered better. 
This is true in other figures as well, please fix everywhere.

Author: val 2-d uses a similar notation to this FYI. Changed this to look cleaner.

  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:gas_yield
  caption=Comparison of TMAP8 total gas-phase hydrogen mass against SRNL experimental cumulative H$_2$ yield data.

> #### Comparison of Steel Hydrogen Uptake Between Models

Reviewer: ```suggestion
#### Conservation of Mass

!style halign=left
[fig:gas_steel_conservation] verifies conservation of mass in the gas-steel model by comparing the total hydrogen mass in the domain (`cylinder_total_mass`, the sum of gas and steel masses) against the sum of the time-integrated boundary flux (`cylinder_time_integrated_flux`) and the cumulative source term (`cylinder_total_generation`). Throughout, all quantities are tracked in $\mathrm{\mu}$mol H atoms: the gas-phase H$_2$ concentration is multiplied by 2 in the `inner_circle_concentration_gas` postprocessor, and the H$_2$ yield function is scaled by `scale_factor = 2` in `cylinder_total_generation`. The two quantities track closely, confirming that mass contributions from the coupled gas-generation, gas-phase transport, interface transfer, and steel diffusion are all consistently accounted for.

!media comparison_mini_canister.py
  image_name=gas_steel_conservation_of_mass.png
  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:gas_steel_conservation
  caption=Conservation of mass check for the gas-steel model: accumulated boundary flux plus source term vs. total H mass in the domain.
  
## Comparison of Steel Hydrogen Uptake Between the Two Models
```

Reviewer: Also, same comment about plotting the ratio rather than the total amounts.

#### Conservation of Mass

!style halign=left
[fig:gas_steel_conservation] verifies conservation of mass in the gas-steel model by comparing the total hydrogen mass in the domain (`cylinder_total_mass`, the sum of gas and steel masses) against the sum of the time-integrated boundary flux (`cylinder_time_integrated_flux`) and the cumulative source term (`cylinder_total_generation`). Throughout, all quantities are tracked in $\mathrm{\mu}$mol H atoms: the gas-phase H$_2$ concentration is multiplied by 2 in the `inner_circle_concentration_gas` postprocessor, and the H$_2$ yield function is scaled by `scale_factor = 2` in `cylinder_total_generation`. The two quantities track closely, confirming that mass contributions from the coupled gas-generation, gas-phase transport, interface transfer, and steel diffusion are all consistently accounted for.

!media comparison_mini_canister.py
  image_name=gas_steel_conservation_of_mass.png
  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:gas_steel_conservation

>   caption=Conservation of mass check for the gas-steel model: accumulated boundary flux plus source term vs. total H mass in the domain.

Reviewer: ```suggestion
```

## Input File Structure

!style halign=left

> Both models are structured around two shared files that are incorporated via the `!include` directive:

Reviewer: ```suggestion
Both models are structured around two shared files that are incorporated via the `!include` capability:
```

# Mini-Canister Hydrogen Transport

TMAP8 is used to model hydrogen transport and permeation through an aluminum-clad used nuclear fuel (AUNF) mini-canister storage device from Savannah River National Laboratory (SRNL) [!citep](d'entremont2024aunfminicanister). The mini-canisters house irradiated AUNF assemblies where gamma and neutron radiation from the fuel drives radiolytic decomposition of water, generating H$_2$ gas. Over time, this hydrogen will disassociate and diffuse through the surrounding 304 stainless steel wall, raising concern for potential accumulation. This example demonstrates how TMAP8 can model these processes through two distinct input files:

> 1. [steel_only.i] — isolates hydrogen diffusion through the steel wall with an assumed boundary partial pressure. This simpler model permits verification against an analytical solution, provided the Dirichlet boundary conditions are time-independent.

Reviewer 2: ```suggestion
1. [steel_only.i] — isolates hydrogen diffusion through the steel wall with an assumed boundary partial pressure. This simpler model permits verification against an analytical solution, with time-independent Dirichlet boundary conditions.
```

2. [gas_steel.i] — simulates the full system: radiolytic H$_2$ generation, gas-phase transport inside the canister, and simultaneous permeation through the steel wall. This model is validated against SRNL experimental measurements [!citep](d'entremont2024aunfminicanister).

Both models share the same 1D axisymmetric geometry and material parameters for the steel wall. The progression from [steel_only.i] to [gas_steel.i] illustrates the flexibility of TMAP8 in building complexity incrementally.

> ## Input File Structure

Reviewer 2: I think it could optimize the structure if you could move this section to input file section at the bottom of this documentation

Each top-level input file adds only the objects that are specific to its model. This structure keeps the shared physics in one place and avoids duplication. Both models simulate 0.25 years (≈ 91.3 days) using a BDF2 time integration scheme with an `IterationAdaptiveDT` adaptive timestep that targets 5 Newton iterations per step.

## Canister Geometry and Mesh

!style halign=left
Both models represent the canister as a 1D axisymmetric domain in cylindrical coordinates ($r$, $z$), using `coord_type = RZ` with `rz_coord_axis = Y` so that the $x$-axis is the radial direction. The canister dimensions [!citep](d'entremont2024aunfminicanister) are listed in [tab:geometry].

!table id=tab:geometry caption=Canister geometry.
| Parameter | Value | Units |
| --- | --- | --- |
| Inner radius, $r_i$ | **1.415 in (35.94 mm)** | mm |
| Steel wall thickness, $\delta_s$ | **0.085 in (2.16 mm)** | mm |
| Outer radius, $r_o = r_i + \delta_s$ | **1.500 in (38.10 mm)** | mm |
| Canister height, $h$ | **7.06 in (179.3 mm)** | mm |
| Internal gas volume, $V_g = \pi r_i^2 h$ | **$\approx$ 727,400** | mm$^3$ |

> In [steel_only.i], the 1D mesh spans only the steel wall, from $r_i$ to $r_o$, using a single `GeneratedMeshGenerator` with 1,500 elements (subdomain 1). In [gas_steel.i], a `CartesianMeshGenerator` produces two adjacent blocks: the gas block (subdomain 0) from $r = 0$ to $r = r_i$ with 250 elements, and the steel block (subdomain 1) from $r = r_i$ to $r = r_o$ with 1,500 elements. Two `SideSetsBetweenSubdomainsGenerator` steps then create the named interface sidesets `interface_gas_to_steel` and `interface_steel_to_gas` that are required for the `ADInterfaceSorption` kernel.

Reviewer 2: ```suggestion
In [steel_only.i], the 1D mesh spans only the steel wall, from $r_i$ to $r_o$, using a single [GeneratedMeshGenerator.md] with 1,500 elements (subdomain 1). In [gas_steel.i], a [CartesianMeshGenerator.md] produces two adjacent blocks: the gas block (subdomain 0) from $r = 0$ to $r = r_i$ with 250 elements, and the steel block (subdomain 1) from $r = r_i$ to $r = r_o$ with 1,500 elements. Two [SideSetsBetweenSubdomainsGenerator.md] steps then create the named interface sidesets `interface_gas_to_steel` and `interface_steel_to_gas` that are required for the [ADInterfaceSorption.md] kernel.
```

> 

Reviewer 2: ```suggestion

In [steel_only.i], the mesh is defined as:
```

!listing test/tests/mini_canister/steel_only.i link=false block=Mesh

> 

Reviewer 2: ```suggestion

In [gas_steel.i], the mesh is defined as:
```

| **$K_s$** | Solubility of H in steel (Sieverts' constant) | $\mu$mol mm$^{-3}$ Pa$^{-1/2}$ |
| **$P$** | Partial pressure of H$_2$ | Pa |
| **$T$** | Temperature (constant) | K |
| **$R$** | Ideal gas constant | J K$^{-1}$ mol$^{-1}$ |
| **$S$** | Volumetric H$_2$ generation rate in gas | $\mu$mol mm$^{-3}$ day$^{-1}$ |
| **$N(t)$** | Cumulative H$_2$ yield in gas | $\mu$mol |
| **$t$** | Time | day |

## Steel-Only Model

### Governing Equations

!style halign=left
In the steel-only model, only hydrogen transport within the steel wall is simulated. The governing equation is a 1D diffusion model:

\begin{equation} \label{eq:steel_diffusion}
\frac{\partial C_s}{\partial t} = \frac{\partial}{\partial r}\left(D_s\frac{\partial C_s}{\partial r} \right),
\end{equation}

where the diffusivity follows an Arrhenius temperature dependence [!citep](san_marchi2012hydrogensteel):

\begin{equation} \label{eq:diffusivity}
D_s(T) = D_{s,0} \exp\!\left( - \frac{E_{D}}{RT} \right).
\end{equation}

Because `coord_type = RZ` is set, MOOSE automatically applies the axisymmetric cylindrical weighting to the diffusion kernel.

### Boundary Conditions

!style halign=left

> At the inner steel surface ($r = r_i$), the hydrogen concentration is fixed by Sieverts' law using the `EquilibriumBC` boundary condition. An auxiliary variable `H_partial_pressure_gas` is first set by a `FunctionAux` at the inner boundary, and `EquilibriumBC` then enforces:

Reviewer 2: ```suggestion
At the inner steel surface ($r = r_i$), the hydrogen concentration is fixed by Sieverts' law using the [EquilibriumBC.md] boundary condition. An auxiliary variable `H_partial_pressure_gas` is first set by a [FunctionAux.md] at the inner boundary, and [EquilibriumBC.md] then enforces:
```
I think including links to these MOOSE features would help readers who are not familiar with MOOSE better understand how the simulation works. Could you also do the same things for other MOOSE features?

\end{equation}

The default configuration uses the constant pressure option:

\begin{equation} \label{eq:assumed_pressure}
P = 0.10 \times 24 \text{ psi} \approx 16{,}547 \text{ Pa},
\end{equation}

which assumes 10% of the 24 psi He-backfilled canister pressure is attributable to H$_2$ [!citep](d'entremont2024aunfminicanister,hlushko2024aunf).

At the outer steel surface ($r = r_o$), hydrogen is released to the ambient environment, and the concentration is set to zero by a Dirichlet condition (defined in [mini_canister_base.i]):

\begin{equation} \label{eq:outer_bc}
C_s(r_o, t) = 0.
\end{equation}

### Solver

!style halign=left
Because the steel-only problem is linear (constant diffusivity, linear Sieverts' BC), `steel_only.i` uses `solve_type = LINEAR` for a direct LU factorization at each timestep, which is more efficient than a nonlinear Newton iteration.

### Model Parameters

!table id=tab:steel_only_params caption=Steel-only model parameters.
| Parameter | Description | Value | Units | Reference |
| --- | --- | --- | --- | --- |
| **$D_{s,0}$** | Diffusivity pre-exponential factor | **$0.20 \times 10^{-6}$** | m$^2$ s$^{-1}$ | [!cite](san_marchi2012hydrogensteel) |

> | **$E_{D}$** | Diffusivity activation energy | **49.3** | kJ mol$^{-1}$ | [!cite](san_marchi2012hydrogensteel) |

Reviewer 2: The bold for "**49.3**" in the table didn't work well at the final documentation.

Reviewer 2: The issue still exist, please just remove the bold symbol for the values in this table and following table to avoid the additional "**" presented in documentation. Below is the table I saw from the compiled documentation in TMAP8 website:

<img width="1034" height="257" alt="Image" src="https://github.com/user-attachments/assets/5a7116ee-c6a6-4552-b562-a644f84f3282" />

| **$K_{s,0}$** | Solubility pre-exponential factor | **$266 \times 10^{-6}$** | $\mu$mol mm$^{-3}$ Pa$^{-1/2}$ | [!cite](san_marchi2012hydrogensteel) |
| **$E_{K}$** | Solubility activation energy | **6.86** | kJ mol$^{-1}$ | [!cite](san_marchi2012hydrogensteel) |
| **$T$** | Temperature | **313.15** | K | [!cite](hlushko2024aunf) |

> | **$P$** | Assumed H$_2$ partial pressure | **$\approx$ 16,547** | Pa | [!cite](hlushko2024aunf) |

Reviewer 2: Same for the bold of these values

### Hydrogen Generation Source Term

!style halign=left
The cumulative radiolytic H$_2$ yield is modeled using a power-law fit to the SRNL experimental data [!citep](d'entremont2024aunfminicanister):

\begin{equation} \label{eq:H2_yield}
N(t) = 69.7055 \, t^{0.6808} \quad [\mu\text{mol}],
\end{equation}

where $t$ is time in days, calibrated assuming a Co-60 irradiator dose rate of approximately 124 Gy/min [!citep](d'entremont2024aunfminicanister). The volumetric source term in [eq:gas_diffusion] is obtained by differentiating [eq:H2_yield] with respect to time and normalizing by the gas volume:

\begin{equation} \label{eq:source_term}
S(t) = \frac{1}{V_g} \frac{dN}{dt} = \frac{69.7055 \times 0.6808}{V_g} \, t^{0.6808 - 1} \quad [\mu\text{mol mm}^{-3}\text{ day}^{-1}].
\end{equation}

### Interface Condition

!style halign=left

> At the gas-steel interface ($r = r_i$), hydrogen equilibrium between the gas and solid phases is enforced by the `ADInterfaceSorption` interface kernel, applied on the `interface_steel_to_gas` sideset. Using the ideal gas law to convert gas-phase concentration to partial pressure ($P = C_g R T$), the equilibrium atomic hydrogen concentration in the steel is:

Reviewer 2: Link may better

\begin{equation} \label{eq:interface_sieverts}
C_s(r_i, t) = 2 K_s(T) \sqrt{C_g(r_i, t) \, R \, T},
\end{equation}

> The `unit_scale_neighbor` parameter is set to $10^3$ to correct for the unit mismatch between $C_g$ in $\mu$mol mm$^{-3}$ and the ideal gas constant used internally by `ADInterfaceSorption`, which draws $R$ in J K$^{-1}$ mol$^{-1}$ from the MOOSE `PhysicalConstants` namespace. Converting $C_g$ from $\mu$mol mm$^{-3}$ to mol m$^{-3}$ introduces a combined factor of $10^{-6}$ (mol/$\mu$mol) $\times$ $10^{9}$ (mm$^3$/m$^3$) $= 10^3$, which is supplied via `unit_scale_neighbor`.

Reviewer 2: Links for ADInterfaceSorption and PhysicalConstants may better

### Solver

!style halign=left
Because the gas-steel problem is nonlinear (the interface couples $C_g$ and $C_s$ through a square-root relationship), `gas_steel.i` uses `solve_type = Newton` with `line_search = NONE`.

### Model Parameters

!table id=tab:gas_steel_params caption=Additional gas-steel model parameters (steel parameters as in [tab:steel_only_params]).
| Parameter | Description | Value | Units | Reference |
| --- | --- | --- | --- | --- |
| **$D_g$** | H$_2$ diffusivity in He | **2.7** | cm$^2$ s$^{-1}$ | [!cite](middha2002hydrogenhelium) |
| **$a$** | H$_2$ yield pre-factor | **69.7055** | $\mu$mol day$^{-0.6808}$ | [!cite](d'entremont2024aunfminicanister) |
| **$b$** | H$_2$ yield exponent | **0.6808** | (dimensionless) | [!cite](d'entremont2024aunfminicanister) |
| **$V_g$** | Internal gas volume | **$\approx$ 727,400** | mm$^3$ | Computed |

> | **$T$** | Temperature | **313.15** | K | [!cite](hlushko2024aunf) |

Reviewer 2: Bold didn't work for values

!media comparison_mini_canister.py
  image_name=gas_steel_conservation_of_mass.png
  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:gas_steel_conservation
  caption=Conservation of mass check for the gas-steel model: accumulated boundary flux plus source term vs. total H mass in the domain.

## Input Files

!style halign=left
The input files for this example are [/steel_only.i] and [/gas_steel.i], together with the shared [/mini_canister_base.i] and [/mini_canister.params]. All files are also used as tests in TMAP8 at [/mini_canister/tests].

!listing test/tests/mini_canister/mini_canister_base.i link=false

!listing test/tests/mini_canister/steel_only.i link=false

!listing test/tests/mini_canister/gas_steel.i link=false

> ## References

Reviewer 2: ```suggestion
```

# Mini-Canister Hydrogen Transport

> TMAP8 is used to model hydrogen transport and permeation through an aluminum-clad used nuclear fuel (AUNF) mini-canister storage device from Savannah River National Laboratory (SRNL) [!citep](d'entremont2024aunfminicanister). The mini-canisters house irradiated AUNF assemblies where gamma and neutron radiation from the fuel drives radiolytic decomposition of water, generating H$_2$ gas. Over time, this hydrogen will disassociate and diffuse through the surrounding 304 stainless steel wall, raising concern for potential accumulation. This example demonstrates how TMAP8 can model these processes through two distinct input files:

Reviewer 2: ```suggestion
TMAP8 is used to model hydrogen transport and permeation through an aluminum-clad used nuclear fuel (AUNF) mini-canister storage device from Savannah River National Laboratory (SRNL) [!citep](d'entremont2024aunfminicanister). The mini-canisters house irradiated AUNF assemblies where gamma and neutron radiation from the fuel drives radiolytic decomposition of water, generating H$_2$ gas. Over time, this hydrogen will dissociate and diffuse through the surrounding 304 stainless steel wall, raising concern for potential accumulation. This example demonstrates how TMAP8 can model these processes through two distinct input files:
```

\end{equation}

!style halign=left
The H$_2$ partial pressure $P$ is provided by a function selected via the `pressure_function` input parameter. Three options are available: (1) a constant value (`constant_pressure`), (2) a [TimeRampFunction.md] from zero to the constant value over 3 hours (`time_ramp_pressure`), or (3) a power-law fit to experimental SRNL pressure data (`SRNL_pressure_data_fun`):

\begin{equation} \label{eq:srnl_pressure}
P_{\text{SRNL}}(t) = 376.7588 \, t^{0.6177} \quad [\text{Pa}].
\end{equation}

The default configuration uses the constant pressure option:

\begin{equation} \label{eq:assumed_pressure}
P = 0.10 \times 24 \text{ psi} \approx 16{,}547 \text{ Pa},
\end{equation}

which assumes 10% of the 24 psi He-backfilled canister pressure is attributable to H$_2$ [!citep](d'entremont2024aunfminicanister,hlushko2024aunf).

At the outer steel surface ($r = r_o$), hydrogen is released to the ambient environment, and the concentration is set to zero by a [DirichletBC.md] (defined in [mini_canister_base.i]):

\begin{equation} \label{eq:outer_bc}
C_s(r_o, t) = 0.
\end{equation}

### Solver

!style halign=left

> Because the steel-only problem is linear (constant diffusivity, linear Sieverts' BC), [steel_only.i] uses 'solve_type = LINEAR' for a direct LU factorization at each timestep.

Reviewer 2: ```suggestion
Because the steel-only problem is linear (constant diffusivity, linear Sieverts' BC), [steel_only.i] uses `solve_type = LINEAR` for a direct LU factorization at each timestep.
```

S(t) = \frac{1}{V_g} \frac{dN}{dt} = \frac{69.7055 \times 0.6808}{V_g} \, t^{0.6808 - 1} \quad [\mu\text{mol mm}^{-3}\text{ day}^{-1}].
\end{equation}

### Interface Condition

!style halign=left
At the gas-steel interface ($r = r_i$), hydrogen equilibrium between the gas and solid phases is enforced by the [InterfaceSorption.md] interface kernel, applied on the `interface_steel_to_gas` sideset. Using the ideal gas law to convert gas-phase concentration to partial pressure ($P = C_g R T$), the equilibrium atomic hydrogen concentration in the steel is:

\begin{equation} \label{eq:interface_sieverts}
C_s(r_i, t) = 2 K_s(T) \sqrt{C_g(r_i, t) \, R \, T},
\end{equation}

The `unit_scale_neighbor` parameter is set to $10^3$ to correct for the unit mismatch between $C_g$ in $\mu$mol mm$^{-3}$ and the ideal gas constant used internally by [InterfaceSorption.md], which draws $R$ in J K$^{-1}$ mol$^{-1}$ from the MOOSE [TMAP8PhysicalConstants.md] namespace. Converting $C_g$ from $\mu$mol mm$^{-3}$ to mol m$^{-3}$ introduces a combined factor of $10^{-6}$ (mol/$\mu$mol) $\times$ $10^{9}$ (mm$^3$/m$^3$) $= 10^3$, which is supplied via `unit_scale_neighbor`.

### Solver

!style halign=left

> Because the gas-steel problem is nonlinear (the interface couples $C_g$ and $C_s$ through a square-root relationship), [gas_steel.i] uses 'solve_type = Newton` with `line_search = NONE`.

Reviewer 2: ```suggestion
Because the gas-steel problem is nonlinear (the interface couples $C_g$ and $C_s$ through a square-root relationship), [gas_steel.i] uses `solve_type = Newton` with `line_search = NONE`.
```

### Model Parameters

!table id=tab:gas_steel_params caption=Additional gas-steel model parameters (steel parameters as in [tab:steel_only_params]).
| Parameter | Description | Value | Units | Reference |
| --- | --- | --- | --- | --- |
| **$D_g$** | H$_2$ diffusivity in He | **2.7** | cm$^2$ s$^{-1}$ | [!cite](middha2002hydrogenhelium) |
| **$a$** | H$_2$ yield pre-factor | **69.7055** | $\mu$mol day$^{-0.6808}$ | [!cite](d'entremont2024aunfminicanister) |

> | **$b$** | H$_2$ yield exponent | **0.6808** | (dimensionless) | [!cite](d'entremont2024aunfminicanister) |

Reviewer 2: $a$ and $b$ is introduce in table but haven been mentioned in documentation. I think you can use symbol of $a$ and $b$ for equation "eq:H2_yield" instead of values, or just remove $a$ and $b$ in this table.

# Mini-Canister Hydrogen Transport

TMAP8 is used to model hydrogen transport and permeation through an aluminum-clad used nuclear fuel (AUNF) mini-canister storage device from Savannah River National Laboratory (SRNL) [!citep](d'entremont2024aunfminicanister). The mini-canisters house irradiated AUNF assemblies where gamma and neutron radiation from the fuel drives radiolytic decomposition of water, generating H$_2$ gas. Over time, this hydrogen will dissociate and diffuse through the surrounding 304 stainless steel wall, raising concern for potential accumulation. This example demonstrates how TMAP8 can model these processes through two distinct input files with varying degrees of fidelity:

1. [steel_only.i] — isolates hydrogen diffusion through the steel wall with an assumed boundary partial pressure. This simpler model permits verification against an analytical solution, assuming time-independent Dirichlet boundary conditions.

> 2. [gas_steel.i] — simulates the full system: radiolytic H$_2$ generation, gas-phase transport inside the canister, and simultaneous permeation through the steel wall. This model is validated against SRNL experimental measurements [!citep](d'entremont2024aunfminicanister).

Reviewer: ```suggestion
2. [gas_steel.i] — simulates the full system: radiolytic H$_2$ generation, gas-phase transport inside the canister, and simultaneous permeation through the steel wall. This model is compared against SRNL experimental measurements [!citep](d'entremont2024aunfminicanister).
```

\begin{equation} \label{eq:srnl_pressure}
P = P_{\text{SRNL}}\left(t\right) \coloneqq 376.7588 \, t^{0.6177} \quad \text{Pa},
\end{equation}

or (2) a constant value (`constant_pressure`) that can be selected to enable the diffusion-front verification against a closed-form analytical solution (see [fig:diffusion_length]):

\begin{equation} \label{eq:assumed_pressure}
P = P_c = 0.10 \times 24 \text{ psi} \approx 16{,}547 \text{ Pa},
\end{equation}

which assumes 10% of the 24 psi He-backfilled canister pressure is attributable to H$_2$ [!citep](d'entremont2024aunfminicanister,hlushko2024aunf).

At the outer steel surface ($r = r_i + \delta$), hydrogen is released to the ambient environment, and the concentration is set to zero by a [DirichletBC.md] (defined in [mini_canister_base.i]):

\begin{equation} \label{eq:outer_bc}
C_s(r_i + \delta, t) = 0.
\end{equation}

### Solver

!style halign=left
Because the steel-only problem is linear (constant diffusivity, linear Sieverts' BC), [steel_only.i] uses `solve_type = LINEAR` for a direct LU factorization at each timestep. The simulation runs for 0.25 years (≈ 91.3 days) using a [BDF2.md] time integration scheme.

### Model Parameters

> 

Reviewer: ```suggestion

[tab:steel_only_params] lists the steel-only model parameters and simulation conditions. 

```
Make sure to cross-reference tables and figures in the text.

| $\mathbf{E_{K}}$ | Solubility activation energy | $\mathbf{6.86}$ | kJ mol$^{-1}$ | [!cite](san_marchi2012hydrogensteel) |
| $\mathbf{T}$ | Temperature | $\mathbf{313.15}$ | K | [!cite](d'entremont2024aunfminicanister) |

### Results

#### Conservation of Mass

!style halign=left
As an internal consistency check, the total hydrogen mass integrated over the steel domain (`annular_cylinder_total_mass_steel`) is compared against the time-integrated net diffusive flux across the inner and outer boundaries (`annular_cylinder_time_integrated_flux`). Both integrals are weighted for the axisymmetric cylindrical geometry and then scaled by the canister height $h$ to represent 3D mass in $\mathrm{\mu}$mol H. [fig:steel_conservation] shows the percent difference between the two quantities, defined as $\left|\,\mathrm{flux}/\mathrm{mass} - 1\,\right| \times 100\,\%$. The metric is highly sensitive to the postprocessors' small values within the first few timesteps, but quickly converges to a small value, and can be further reduced by refining the mesh and timestep.

> !media comparison_mini_canister.py

Reviewer: I remember the mass conservation error being much smaller than this. Could you reduce it further?

Author: This is a numerical issue. The relative metric is quite sensitive to the small flux and mass values early in the simulations. Refining the initial starting timestep smaller than 300 seconds can lead to negative concentrations if my memory serves, and I do not want to employ the bounds system if we don't have to.

The analytical diffusion length is 0.175 mm for a 0.25 year simulation, where the steel thickness is 2.159 mm. So most of the elements remain unchanged. To keep the simulations running quickly for tests, I implemented a bias that shrinks element size near the interface of the gas and steel for the steel_only simulation. This also improves the conservation of mass metrics. I did not implement this change for the gas-steel model as that will require a new meshing strategy entirely. I would expect the conservation metrics to match those of the steel-only case more closely, but the error values are acceptable. Let me know if you disagree and want me to change the mesh to match between the two models.

I just pushed a change that calculates the absolute metric, total unaccounted mass in micromols |flux - mass|, until some specified threshold of hydrogen accumulation. Early timesteps see an absolute error that is sub-micromol. After this threshold is achieved, the code calculates the relative metric for percent difference.

!media comparison_mini_canister.py
  image_name=gas_steel_conservation_of_mass.png
  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:gas_steel_conservation
  caption=Conservation of mass for the gas-steel model: percent difference $\left|\,(\mathrm{flux}+\mathrm{source})/\mathrm{mass} - 1\,\right| \times 100\,\%$ between the accumulated boundary flux plus source term and the total H mass in the domain for $t>1$ day.

#### Gas-Phase Hydrogen Yield Calculations

!style halign=left

> [fig:gas_yield] compares the total atomic hydrogen mass in the gas phase (`inner_cylinder_total_mass_gas`) against the cumulative H$_2$ yield measured by SRNL [!citep](d'entremont2024aunfminicanister). Agreement between the simulation and experiment reflects the accuracy of the power-law source model ([eq:H2_yield]). It is important to note that the simulation results are compared against data that is fed into the model. Future work will include a more complex generation model independent of this data.

Reviewer: ```suggestion
[fig:gas_yield] compares the total atomic hydrogen mass in the gas phase (`inner_cylinder_total_mass_gas`) against the cumulative H$_2$ yield measured by SRNL [!citep](d'entremont2024aunfminicanister). Agreement between the simulation and experiment reflects the accuracy of the power-law source model ([eq:H2_yield]). It is important to note that the simulation results are compared against data that is fed into the model, which does not enable this effort to be considered a validation. Future work will include a more complex generation model independent of this data.
```

\end{equation}

!style halign=left
The H$_2$ partial pressure $P$ is provided by a function selected via the `pressure_function` input parameter. A nonexhaustive set of implementations include: (1) a power-law fit to experimental SRNL pressure data (`SRNL_pressure`) calculated using a power-law least-squares fit, used as the default for validation (see [fig:partial_pressure] and [fig:gas_yield]) and for the conservation-of-mass check (see [fig:steel_conservation]):

\begin{equation} \label{eq:srnl_pressure}
P = P_{\text{SRNL}}\left(t\right) \coloneqq 376.7588 \, t^{0.6177} \quad \text{Pa},
\end{equation}

or (2) a constant value (`constant_pressure`) that can be selected to enable the diffusion-front verification against a closed-form analytical solution (see [fig:diffusion_length]):

\begin{equation} \label{eq:assumed_pressure}

> P = P_c \coloneqq 0.10 \times 24 \text{ psi} \approx 16{,}547 \text{ Pa},

Reviewer: ```suggestion
P = P_c = 0.10 \times 24 \text{ psi} \approx 16{,}547 \text{ Pa},
```

\end{equation}

### Solver

!style halign=left
Because the steel-only problem is linear (constant diffusivity, linear Sieverts' BC), [steel_only.i] uses `solve_type = LINEAR` for a direct LU factorization at each timestep.

### Model Parameters and Simulation Conditions

!table id=tab:steel_only_params caption=Steel-only model parameters and simulation conditions.
| Parameter | Description | Value | Units | Reference |
| --- | --- | --- | --- | --- |
| $\mathbf{D_{s,0}}$ | Diffusivity pre-exponential factor | $\mathbf{0.20 \times 10^{-6}}$ | m$^2$ s$^{-1}$ | [!cite](san_marchi2012hydrogensteel) |
| $\mathbf{E_{D}}$ | Diffusivity activation energy | $\mathbf{49.3}$ | kJ mol$^{-1}$ | [!cite](san_marchi2012hydrogensteel) |
| $\mathbf{K_{s,0}}$ | Solubility pre-exponential factor | $\mathbf{266 \times 10^{-6}}$ | $\mathrm{\mu}$mol mm$^{-3}$ Pa$^{-1/2}$ | [!cite](san_marchi2012hydrogensteel) |
| $\mathbf{E_{K}}$ | Solubility activation energy | $\mathbf{6.86}$ | kJ mol$^{-1}$ | [!cite](san_marchi2012hydrogensteel) |
| $\mathbf{T}$ | Temperature | $\mathbf{313.15}$ | K | [!cite](d'entremont2024aunfminicanister) |

### Results

#### Conservation of Mass

!style halign=left

> As an internal consistency check, the total hydrogen mass integrated over the steel domain (`annular_cylinder_total_mass_steel`) is compared against the time-integrated net diffusive flux across the inner and outer boundaries (`annular_cylinder_time_integrated_flux`). Both integrals are weighted for the axisymmetric cylindrical geometry and then scaled by the canister height $h$ to represent 3D mass in $\mathrm{\mu}$mol H. [fig:steel_conservation] shows the percent difference between the two quantities as a single-curve conservation metric, defined as $\left|\,\mathrm{flux}/\mathrm{mass} - 1\,\right| \times 100\,\%$. The metric is highly sensitive to the postprocessors' small values within the first few timesteps, but quickly converges to a small value, and can be further reduced by refining the mesh and timestep.

Reviewer: ``` suggestion
As an internal consistency check, the total hydrogen mass integrated over the steel domain (`annular_cylinder_total_mass_steel`) is compared against the time-integrated net diffusive flux across the inner and outer boundaries (`annular_cylinder_time_integrated_flux`). Both integrals are weighted for the axisymmetric cylindrical geometry and then scaled by the canister height $h$ to represent 3D mass in $\mathrm{\mu}$mol H. [fig:steel_conservation] shows the percent difference between the two quantities, defined as $\left|\,\mathrm{flux}/\mathrm{mass} - 1\,\right| \times 100\,\%$. The metric is highly sensitive to the postprocessors' small values within the first few timesteps, but quickly converges to a small value, and can be further reduced by refining the mesh and timestep.
```

> 

Reviewer: ```suggestion

In the input files, these two postprocessors are defined as:

!listing test/tests/mini_canister/steel_only.i link=false block=Postprocessors/exact_diffusion_length

!listing test/tests/mini_canister/steel_only.i link=false block=Postprocessors/simulated_diffusion_length

```

!media comparison_mini_canister.py
  image_name=steel_only_conservation_of_mass.png
  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:steel_conservation

>   caption=Conservation of mass check for the steel-only model: percent difference $\left|\,\mathrm{flux}/\mathrm{mass} - 1\,\right| \times 100\,\%$ between the accumulated boundary flux and the total H mass in the steel domain for $t>1$ day.

Reviewer: ```suggestion
Conservation of mass for the steel-only model: percent difference $\left|\,\mathrm{flux}/\mathrm{mass} - 1\,\right| \times 100\,\%$ between the accumulated boundary flux and the total H mass in the steel domain for $t>1$ day.
```

#### Diffusion Front Verification

!style halign=left
For a semi-infinite slab with a constant-concentration boundary condition, the diffusion front advances as $\ell(t) = \sqrt{\pi D_s t}$, providing a straightforward check that the numerical diffusion is correctly implemented. The simulated diffusion front is computed via the `simulated_diffusion_length` postprocessor as the $x$-intercept of the tangent line using the interface concentration and gradient. [fig:diffusion_length] shows the simulated diffusion front length, using the constant pressure $P_c$, compared to the analytical expression, which it matches.

!media comparison_mini_canister.py
  image_name=diffusion_length.png
  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:diffusion_length
  caption=Comparison of the simulated and analytical ($\sqrt{\pi D_s t}$) diffusion front length in the steel wall over 0.25 years.

!listing test/tests/mini_canister/steel_only.i link=false block=Postprocessors/exact_diffusion_length

> !listing test/tests/mini_canister/steel_only.i link=false block=Postprocessors/simulated_diffusion_length

Reviewer: ```suggestion
```

\frac{\partial C_g}{\partial t} = \frac{\partial}{\partial x}\left(D_g \frac{\partial C_g}{\partial x} \right) + S(t),
\end{equation}

> where $D_g$ is the diffusivity of H$_2$ in the He backfill gas [!citep](middha2002hydrogenhelium). In the steel wall ($r_i \leq r \leq r_i + \delta$), [eq:steel_diffusion] applies as before. The axisymmetric weighting is again applied automatically by MOOSE. At the symmetry axis ($r = 0$), no explicit boundary condition is required: MOOSE's natural boundary condition enforces zero diffusive flux.

Reviewer: ```suggestion
where $D_g$ is the diffusivity of H$_2$ in the He backfill gas [!citep](middha2002hydrogenhelium). In the steel wall ($r_i \leq r \leq r_i + \delta$), [eq:steel_diffusion] applies as before. The axisymmetric weighting is again applied automatically by MOOSE. At the symmetry axis ($r = 0$), MOOSE's natural boundary condition enforces zero diffusive flux.
```

\end{equation}

### Interface Condition

!style halign=left
At the gas-steel interface ($r = r_i$), hydrogen equilibrium between the gas and solid phases is enforced by the [InterfaceSorption.md] interface kernel. Using the ideal gas law to convert gas-phase concentration to partial pressure ($P = C_g R T$), the equilibrium atomic hydrogen concentration in the steel is:

\begin{equation} \label{eq:interface_sieverts}
C_s(r_i, t) = 2 K_s(T) \sqrt{C_g(r_i, t) \, R \, T}.
\end{equation}

The `unit_scale_neighbor` parameter is set to $10^3$ to correct for the unit mismatch between $C_g$ in $\mathrm{\mu}$mol mm$^{-3}$ and the ideal gas constant used internally by [InterfaceSorption.md], which draws $R$ in J K$^{-1}$ mol$^{-1}$ from the MOOSE [TMAP8PhysicalConstants.md] namespace. Converting $C_g$ from $\mathrm{\mu}$mol mm$^{-3}$ to mol m$^{-3}$ introduces a combined factor of $10^{-6}$ (mol/$\mathrm{\mu}$mol) $\times$ $10^{9}$ (mm$^3$/m$^3$) $= 10^3$, which is supplied via `unit_scale_neighbor`. The extra factor of $2$ converts the solubility to represent atomic $H$.

### Solver

!style halign=left

> Because the interface is nonlinear, [gas_steel.i] uses `solve_type = Newton`. Both models simulate 0.25 years (≈ 91.3 days) using a [BDF2.md] time integration scheme. This model is equipped with an [IterationAdaptiveDT.md] adaptive timestep that targets 5 Newton iterations per step.

Reviewer: Do not mention `both models` here. You should be focusing on the model being discussed. If it is true for the other model as well, then specify that there too.

### Results

#### Conservation of Mass

!style halign=left
[fig:gas_steel_conservation] verifies conservation of mass in the gas-steel model by comparing the total hydrogen mass in the domain (`cylinder_total_mass`, the sum of gas and steel masses) against the sum of the time-integrated boundary flux (`cylinder_time_integrated_flux`) and the cumulative source term (`cylinder_total_generation`). Throughout, all quantities are tracked in $\mathrm{\mu}$mol H atoms. The figure plots the percent difference between the two quantities, defined similarly to before as $\left|\,(\mathrm{flux}+\mathrm{source})/\mathrm{mass} - 1\,\right| \times 100\,\%$. The metric remains small, confirming that mass contributions from the coupled gas-generation, gas-phase transport, interface transfer, and steel diffusion are all consistently accounted for.

!media comparison_mini_canister.py
  image_name=gas_steel_conservation_of_mass.png
  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:gas_steel_conservation

>   caption=Conservation of mass check for the gas-steel model: percent difference $\left|\,(\mathrm{flux}+\mathrm{source})/\mathrm{mass} - 1\,\right| \times 100\,\%$ between the accumulated boundary flux plus source term and the total H mass in the domain for $t>1$ day.

Reviewer: ```suggestion
  caption=Conservation of mass for the gas-steel model: percent difference $\left|\,(\mathrm{flux}+\mathrm{source})/\mathrm{mass} - 1\,\right| \times 100\,\%$ between the accumulated boundary flux plus source term and the total H mass in the domain for $t>1$ day.
```

> #### Gas-Phase Hydrogen Yield Validation

Reviewer: ```suggestion
#### Gas-Phase Hydrogen Yield Calculations
```

  id=fig:gas_yield
  caption=Comparison of TMAP8 total gas-phase hydrogen mass against SRNL experimental cumulative H$_2$ yield data.

> #### Partial Pressure Validation

Reviewer: ```suggestion
#### Partial Pressure Calculations
```
It needs to be very clear that this is NOT a validation.

> #### Partial Pressure Validation

Reviewer: Can you explain what this section brings in addition to the previous results?

  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:partial_pressure
  caption=Comparison of TMAP8 simulated H$_2$ partial pressure at the gas-steel interface against SRNL experimental measurements.

## Comparison of Steel Hydrogen Uptake Between the Two Models

!style halign=left
[fig:model_comparison] compares the total atomic hydrogen mass accumulated in the steel wall between the steel-only and gas-steel models. The left axis shows the absolute H mass in $\mathrm{\mu}$mol; the right axis shows the steel mass as a percentage of the total H inventory. By default the steel-only model uses the time-dependent SRNL partial-pressure fit (`pressure_function = SRNL_pressure`), giving decent agreement with the gas-steel model, which evolves the interface pressure self-consistently.

!media comparison_mini_canister.py
  image_name=hydrogen_yield_in_steel.png
  style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
  id=fig:model_comparison
  caption=Comparison of total H mass in the steel wall (left axis) and fraction of total H inventory in the steel (right axis, dashed) between the steel-only (SRNL pressure fit) and gas-steel simulations.

> 

Reviewer: ``` suggestion

The conclusion of this analysis is that only a small fraction of hydrogen permeates into the steel canister over the relevant time frame in these conditions. This observation is supported by both models. 
```
